src/server/managers/AppExternalComponentManager.ts

```diff
@@ -1,79 +1,119 @@
-     * or enabled. As long as an app touched an external component, then
-     * it is listed here.
+     * or enabled. As long as an app touched external components, then
+     * they are listed here.
      */
-    private appTouchedComponents: Map<string, IExternalComponent>;
+    private appTouchedExternalComponents: Map<string, Map<string, IExternalComponent>>;
 
     constructor() {
-        this.providedComponents = new Map<string, IExternalComponent>();
-        this.appTouchedComponents = new Map<string, IExternalComponent>();
+        this.registeredExternalComponents = new Map<string, Map<string, IExternalComponent>>();
+        this.appTouchedExternalComponents = new Map<string, Map<string, IExternalComponent>>();
     }
     /**
-     * Add the external component to the appTouchedComponents.
+     * Get all registered components.
+     */
+    public getRegisteredExternalComponents(): Map<string, Map<string, IExternalComponent>> {
+        return this.registeredExternalComponents;
+    }
+    /**
+     * Get all external components that apps have registered
+     * before, including disabled apps' external components.
+     */
+    public getAppTouchedExternalComponents(): Map<string, Map<string, IExternalComponent>> {
+        return this.appTouchedExternalComponents;
+    }
+    /**
+     * Get all external components of an app by specifying the appId.
      *
      * @param appId the id of the app
-     * @param externalComponent the external component to register
-     * @param addToProvidedComponents whether also add the external
-     * component to the providedComponents, the value is false by default.
      */
-    public registerComponent(appId: string, externalComponent: IExternalComponent, addToProvidedComponents: boolean = false): void {
-        this.appTouchedComponents.set(appId, externalComponent);
-
-        if (addToProvidedComponents) {
-            this.providedComponents.set(appId, externalComponent);
+    public getExternalComponents(appId: string): Map<string, IExternalComponent> {
+        if (this.appTouchedExternalComponents.has(appId)) {
+            return this.appTouchedExternalComponents.get(appId);
         }
+
+        return null;
     }
     /**
-     * Remove the external component from the providedComponents by appId.
+     * Add an external component to the appTouchedExternalComponents.
+     * If you call this method twice and the component
+     * has the same name as before, the first one will be
+     * overwritten as the names provided **must** be unique.
      *
      * @param appId the id of the app
+     * @param externalComponent the external component need to be added
      */
-    public unregisterComponent(appId: string): void {
-        if (this.providedComponents.has(appId)) {
-            this.providedComponents.delete(appId);
+    public addExternalComponent(appId: string, externalComponent: IExternalComponent): void {
+        if (externalComponent.appId !== appId) {
```

Comment:
@lolimay what would be the case this happens? Would it be that an app forgot to include it? Or they provided the wrong one? Maybe we should just set it for them instead and make things simpler?

Reply:
But if we don't add this detection, how can we ensure the develooer pass another app's appId to deceive apps-engine?

Reply:
This exact method is provided the caller's app id by the engine system (https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/145/files#diff-db7e7559fd4be39c41521a0e1b838863R10). So, that is why we can trust the `appId` provided and assign it to the `externalComponent.appId` for the App. That would ensure the external component always has the app id of whichever app provided it.

Reply:
In the current implementation, the app's developer can specify the appId of the external component freely (https://github.com/lolimay/frvr-games-rocketchat-wrapper/blob/99dae2e5fcd428ce41c945d311cc55f3653de53d/basketball-beta/BasketballBetaApp.ts#L17-L26).
Maybe we need also change it to avoid app developers' specifying the appId?

Reply:
No, I wouldn't go that far since it would require separate models for having the `appId` limited. I would just go ahead and overwrite it inside of this method.

Reply:
Got it. Will change it as suggested 👍

Reply:
Resolved in https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/150/commits/26c638d51f1838e5ec6a971ad9409cc9df17e829.